Re-write balanceTransaction using Node.evaluateTransactionBalance

Anviking opened the pull request:
- [x] Rewrite balanceTransaction using `evaluateTransactionBalance`
     - Effects on goldens can be viewed in the diff of this PR
- [x] Enable `prop_balanceTransactionBalanced`
- [x] Polish
- [x] Even more polish
    - [x] Will look through from top to bottom before merging

### Future work
- [x] Pay surplus as excess fee when we cannot afford a change-output. (Would like to do while comparing constructTx with balanceTx)
- [x] Ensure tx size limit is respected (Requires fixing ADP-1484 _and_ [this](https://github.com/input-output-hk/cardano-wallet/blob/ae8a2206b6f659b879d3bf14fd4ffdad7e781992/lib/core/src/Cardano/Wallet.hs#L1788-L1818) )
 
### Comments

First part introducing goldens and the evaluateTransactionBalance function: #3150

### Issue Number


ADP-1372

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

Anviking committed:
Add shrinker for fee

Anviking committed:
Rewrite balanceTransaction using evalauteTransactionBalance

The core idea is to 1) evaluate the balance of the partial tx using the
already existing `evaluateTransactionBalance` and then 2) ask
coin-selection to select enough inputs to cover it.

The key pieces of code are:

```
(balance0, fee0) <- balanceAfterSettingMinFee partialTx
```

```
    { assetsToMint = positiveTokens <> tokensInOutputs
    , assetsToBurn = negativeTokens <> tokensInInputs
    , extraCoinIn =
        positiveAda
        <> adaInOutputs
        <> fromCardanoLovelace fee0
    , extraCoinOut = negativeAda <> adaInInputs
    }
```

The effects are:
1. Can now balance transactions containing deposits and reclaims (e.g.
  first time delegation txs)
2. Enabling of the `prop_balanceTransactionBalanced` property
3. Fixes of several variable-length coin-encoding related corner cases,
   which previously would have resulted in internal server errors or
   improperly balanced results.

Note that the maximum transaction size is not respsected in this commit.
Future commits will fix that.

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1752,6 +1701,180 @@ balanceTransaction
+                        (view #computeSelectionLimit tl)
+                            pp'
+                            defaultTransactionCtx
+                            outputs
```
This is the part I'm uncertain about... Can this be merged as is?

Testing that `balanceTransaction` doesn't create too large txs would require fixing [ADP-1484](https://input-output.atlassian.net/browse/ADP-1484). I would not want to make this PR dependent on that.

Testing that `balanceTransaction` doesn't fail needlessly with `SelectionLimitReached` might be possible as outlined by the comment, but it might take a few days.

Perhaps the
```
let constraints = txConstraints pparams tag
                     --     nWits = estimateNumberOfWitnesses tx
                     --     budget = getTxMaxSize
                     --          - serializedSize tx
                     --          - nWits * (sizeOfWitness constraints)
                     --  in
                     --      budget `div` (sizeOfInput constraints)
```
approach suggested above, but without testing is good enough (for now). Misjudging what we can fit in a tx by a few bytes shouldn't be the end of the world, given that we already may misjudge because of change (ADP-1484)

Anviking replied:
N/A thanks to new approach supported by #3164

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1503,45 +1506,26 @@ balanceTransaction
-                        SelectionCollateralNotRequired
-                } & padFeeEstimation partialTx
-
-        -- FIXME: The coin selection and reported fees will likely be wrong in
-        -- the presence of certificates (and deposits / refunds). An immediate
-        -- "fix" is to return a proper error from the handler when any key or
-        -- pool registration (resp. deregistration) certificate is found in the
-        -- transaction. A long-term fix is to handle this case properly during
-        -- balancing.
```
🎉

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1458,7 +1467,6 @@ instance Buildable PartialTx where
 balanceTransaction
     :: forall m s k ctx.
         ( HasTransactionLayer k ctx
-        , HasLogger m WalletWorkerLog ctx
```
Should probably preserve the logging in the new `selectAssets'`

Anviking replied:
Should be easily doable to preserve: https://github.com/input-output-hk/cardano-wallet/blob/70cc0619e7c23def2747ab602ef5602741a61076/lib/core/src/Cardano/Wallet.hs#L2144-L2149

Possibly some new log message could be added as well… Should look at it…

Anviking replied:
I added a `MsgSelectionForBalancingStart UTxO PartialTx` analogous to `MsgSelectionStart UTxO [TxOut]`, and kept the SelectionError/Report ones the same.

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs:
```diff
@@ -1803,17 +1803,13 @@ spec = describe "NEW_SHELLEY_TRANSACTIONS" $ do
+        -- PlutusScenario.pingPong_1 is sending out 2₳
+        let amt = 2_000_000
         let balancePayload = Json PlutusScenario.pingPong_1
-        let hasExpectedFee = [expectField (#fee . #getQuantity) (`shouldBe` expectedFee)]
```
Could revisit as part of this PR 🤔 

When I was talking to Piotr I didn't know paying higher fees when we cannot afford a change output was already the normal behaviour of the old tx workflow.

Anviking replied:
Supporting higher fees seemed to fix the pending tests without many changes 👍 

However, I'd now like to be clearer about:
- Why I needed to tweak the wallet balance in `TRANS_NEW_BALANCE_02c`
- Why there is a slight regression for balancing pingPong_2: wallet now requires 3 ada rather than 2.75 ada

Maybe improving logging will give some insight.

Anviking replied:
> Why I needed to tweak the wallet balance in TRANS_NEW_BALANCE_02c
> Why there is a slight regression for balancing pingPong_2: wallet now requires 3 ada rather than 2.75 ada

After looking at it, the cases for 3 and 3.05 ada make sense to me: https://github.com/input-output-hk/cardano-wallet/pull/3100#discussion_r818642803

I'm not sure what changed, but it might be that the new behaviour is more "correct" as for being conservative, and it's such a small change in behaviour that I wouldn't worry about it.

Anviking commented on lib/shelley/test/data/balanceTx/pingPong_2/golden:
```diff
@@ -53,12 +53,12 @@
+ 2.850000,ErrBalanceTxSelectAssets (ErrSelectAssetsSelectionError (SelectionCollateralErrorOf (SelectionCollateralError {largestCombinationAvailable = fromList [((TxIn {inputId = Hash "00000000000000000000000000000000", inputIx = 0},Address "`\177\229\224\251t\200l\128\USdhA\224|\219B\223\139\130\239<\228\229|\181A.w"),Coin 2850000)], minimumSelectionAmount = Coin 4275000})))
+ 2.900000,ErrBalanceTxSelectAssets (ErrSelectAssetsSelectionError (SelectionCollateralErrorOf (SelectionCollateralError {largestCombinationAvailable = fromList [((TxIn {inputId = Hash "00000000000000000000000000000000", inputIx = 0},Address "`\177\229\224\251t\200l\128\USdhA\224|\219B\223\139\130\239<\228\229|\181A.w"),Coin 2900000)], minimumSelectionAmount = Coin 4350000})))
+ 2.950000,ErrBalanceTxSelectAssets (ErrSelectAssetsSelectionError (SelectionCollateralErrorOf (SelectionCollateralError {largestCombinationAvailable = fromList [((TxIn {inputId = Hash "00000000000000000000000000000000", inputIx = 0},Address "`\177\229\224\251t\200l\128\USdhA\224|\219B\223\139\130\239<\228\229|\181A.w"),Coin 2950000)], minimumSelectionAmount = Coin 4425000})))
+ 3.000000,ErrBalanceTxSelectAssets (ErrSelectAssetsSelectionError (SelectionCollateralErrorOf (SelectionCollateralError {largestCombinationAvailable = fromList [((TxIn {inputId = Hash "00000000000000000000000000000000", inputIx = 0},Address "`\177\229\224\251t\200l\128\USdhA\224|\219B\223\139\130\239<\228\229|\181A.w"),Coin 3000000)], minimumSelectionAmount = Coin 4500000})))
```
This may appear odd. When we have 3 ada we get an error saying that we need at least 4.5 ada. But with 3.05 ada, balancing succeeds.

The reason is:
1. The minimum fee is around 2 ada (≈0.5 from the size, and ≈1.5 for the _maximum_ ex unit costs)
2. With an excess of _just_ 1 ada, this is right at the limit of where coin-selection can't create a change output. Instead it returns a higher "selection delta" / "fee".
3. The collateral requirement is based on the actual `fee`, not the `minfee`. With `collateralPercentage: 150%`, this results in a 3 ada *1.5 = 4.5 ada collateral requirement.

Anviking committed:
Re-add logging for balanceTransaction

Anviking committed:
Add guardTxSize to balanceTransaction

This uses `estimateSignedTxSize` to verify that the final result is not
bigger than allowed.

Anviking committed:
Implement SelectionStrategyMinimal fallback in balanceTransaction

If `SelectionStrategyOptimal` results in the tx max size being exceeded,
try again with `SelectionStrategyMinimal`.

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs:
```diff
@@ -1805,9 +1805,6 @@ spec = describe "NEW_SHELLEY_TRANSACTIONS" $ do
     it "TRANS_NEW_BALANCE_05/ADP-1286 - \
         \I can balance correctly in case I need to spend my remaining ADA for fee" $
         \ctx -> runResourceT $ do
-        liftIO $ pendingWith
-            "ADP-1286 - ValueNotConservedUTxO: Transaction seems balanced incorrectly \
-            \in case when less than minUtxOValue is left on the wallet"
```
PR will fix https://input-output.atlassian.net/browse/ADP-1286

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1780,6 +1814,140 @@ balanceTransaction
+            in
+                flip evalRand (stdGenFromSeed seed)
+                    $ runExceptT
+                    $ performSelection selectionConstraints selectionParams
```
This is the idea discussed in #3093

Anviking committed:
Add txMaxSize goldens for balanceTransaction

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1780,6 +1814,140 @@ balanceTransaction
+                        SelectionCollateralRequired
+                    else
+                        SelectionCollateralNotRequired
+
```
```suggestion
```

Anviking commented on lib/core/src/Cardano/Api/Gen.hs:
```diff
@@ -256,8 +256,14 @@ genLovelace = Lovelace <$> frequency
+        offset <- frequency
+            [ (1, choose (-10, 10))
+            , (1, choose (-220_000, -150_000))
+            , (1, choose (150_000, 220_000))
```
Tried to find corner-cases more consistently by adding coin values close to `boundary ± typicalFee`.

Anviking committed:
fixup duplicate error

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1634,6 +1634,48 @@ balanceTransaction
+        -- minUTxOValue of ≈1 ada. Then we only need to care about the following
+        -- two sizes of CBOR lovelace values:
+        -- - 0.065536 ada is encoded as 5 bytes
+        -- - 4.294967296 ada is encoded as 9 bytes
```
clarify that this is the lower bound

Anviking replied:
done

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1500,28 +1508,23 @@ balanceTransaction
 
-    (delta, extraInputs, extraCollateral, extraOutputs) <- do
+    -- TODO: Nothing guarantees consistency between externalInputs and the
+    -- actual txins!
```
Create a task for this

Anviking replied:
Think through

jonathanknowles replied:
@Anviking I agree it's a good idea to make a task for this.

Do you think it qualifies as "technical debt"?

Regardless, perhaps after creating a task, we can make a "TODO" item in the code that references the task.

What do you think?

Anviking replied:
:+1: 2ed7982740998eae1aca0a25701043389441daef and https://input-output.atlassian.net/browse/ADP-1544

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1589,32 +1572,49 @@ balanceTransaction
         , extraCollateral
         , extraOutputs
-        , feeUpdate = UseNewTxFee delta
+        , feeUpdate = UseNewTxFee $ unsafeFromLovelace minfee0
```
Could think about the effect of the size of the fee field

Anviking replied:
I remember now. What makes this work is the assumption that the fee with the current mainnet parameters always is encoded with 5 bytes.

Relates to https://input-output.atlassian.net/browse/ADP-1514 anyhow: c2a3eddc3

We _could_ revert back to using `delta`... but I'd rather keep it like this until tackling ADP-1514

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1589,32 +1572,49 @@ balanceTransaction
+    -- change output with the appropriate minUTxOValue.
+    let extraFeeToBurn =
+            case extraOutputs of
+                [] | surplus > Coin 20_000_000 -> error $ unwords
```
Define in terms of `minUTxOValue`

Anviking replied:
This is not entirely straight-forward. The `minUTxOValue` isn't a protocol parameter anymore in Alonzo. 

One could perhaps use this function https://github.com/input-output-hk/cardano-wallet/blob/8106b842c80c164664030da879308ef018c33f9a/lib/core/src/Cardano/Wallet/Primitive/Types/Tx.hs#L891 if we pass in a `TokenMap.empty`, but without knowing for sure that it's 100% reliable and works in this context, the 20 ada safety limit seems safer.

Anviking commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2541,8 +2534,8 @@ prop_balanceTransactionBalanced
     -> StdGenSeed
     -> Property
-prop_balanceTransactionBalanced (Wallet' utxo wal pending) (ShowBuildable partialTx) seed
-    = withMaxSuccess 200 $ do
```
Could rename `Balanced` -> `Valid`

Anviking committed:
Fixup: add Eq and Show

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1780,6 +1813,140 @@ balanceTransaction
+                -- We need to compensate for inputs and outputs affecting both
+                -- 1) the @balance@ itself and coin-selection and 2) the coin
+                -- selections notion of balance. Hence we add @-sum inputs + sum
+                -- outputs@ to cancel the effect of (2).
```
```suggestion
```

Redundant with comment just above

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -3594,3 +3767,38 @@ instance HasSeverityAnnotation TxSubmitLog where
+-- | Convert a 'Cardano.Value' into a positive and negative component. Useful
+-- to convert the potentially negative balance of a partial tx into
+-- TokenBundles.
+posAndNegFromCardanoValue
```
This function is essential to the function of `balanceTransaction` and is covered indirectly by `prop_balanceTransactionBalanced` and similar.

But I see there's an existing `toCardanoValue :: TokenBundle -> Cardano.Value` function. So a test like
```haskell
forAll \(v :: Cardano.Value) -> do
    let (pos, neg) = posAndNegFromCardanoValue 
    (toCardanoValue pos <> toCardanoValue neg) === v
```
might be straight-forward.

I don't think it's essential at _this stage_,  but I could give it a shot when I have time.

Anviking replied:
Also: I'm imagining we'd replace the `(TokenBundle.TokenBundle, TokenBundle.TokenBundle)` in the future with something that can represent negative quantities (like `Cardano.Value`) (in coin-selection)

jonathanknowles replied:
I'm guessing the round-trip test would need to include a negation:

```hs
forAll \(v :: Cardano.Value) -> do
    let (pos, neg) = posAndNegFromCardanoValue 
    (toCardanoValue pos <> (negate <$> toCardanoValue neg)) === v
```

jonathanknowles replied:
> Also: I'm imagining we'd replace the (TokenBundle.TokenBundle, TokenBundle.TokenBundle) in the future with something that can represent negative quantities (like Cardano.Value) (in coin-selection)

Once coin selection no longer depends on `TokenBundle` and `TokenMap`, then we can potentially remove these from the wallet in favour of using Cardano API types.

This is part of the scope of https://input-output.atlassian.net/browse/ADP-1449.

Anviking replied:
The property passes, reassuringly: 7de15e6566bfb216f0860e954418f52999061c03

jonathanknowles commented on lib/core/src/Cardano/Api/Gen.hs:
```diff
@@ -256,8 +256,14 @@ genLovelace = Lovelace <$> frequency
+            [ (1, choose (-10, 10))
+            , (1, choose (-220_000, -150_000))
+            , (1, choose (150_000, 220_000))
+            , (1, choose (-1_000_000, 1_000_000))
```
@Anviking I think these constants would benefit from a comment explaining why they were chosen.

What do you think?

Anviking replied:
Done

jonathanknowles committed:
Fix a few formatting issues.

- Wrap very long lines.
- Use 4 character indents.

https://input-output-hk.github.io/adrestia/code/Coding-Standards#limit-line-length-to-80-characters

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1484,44 +1494,65 @@ instance Buildable PartialTx where
-    (PartialTx partialTx@(cardanoTx -> Cardano.InAnyCardanoEra _ (Cardano.Tx (Cardano.TxBody bod) _)) externalInputs redeemers)
+    (internalUtxoAvailable, wallet, _pendingTxs)
+    selectionStrategy
+    ptx@(PartialTx partialTx@(cardanoTx -> Cardano.InAnyCardanoEra _ (Cardano.Tx (Cardano.TxBody bod) _)) externalInputs redeemers)
```
This could benefit from some wrapping, as:
- it's hard to see where the nesting is
- it's a very long line

I've made a commit here which fixes the issue:
fb01b2bb95e146e0d6f20b3a558e3fd4085edf3f

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1484,44 +1494,65 @@ instance Buildable PartialTx where
+            ErrBalanceTxMaxSizeLimitExceeded
+                -> balanceWith SelectionStrategyMinimal
+            otherErr
+                -> throwE otherErr
```
👍🏻

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -2007,10 +2174,10 @@ selectAssets ctx pp params transform = do
             , extraCoinIn = Coin 0
             , extraCoinOut = Coin 0
+            , assetsToBurn =
+                params ^. (#txContext . #txAssetsToBurn)
```
Any reason why this one gets moved, but not `assetsToMint`? 😄

Anviking replied:
Oops. Might have happened during a rebase. Reverted to the original location.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2535,33 +2553,33 @@ prop_balanceTransactionBalanced
+                    $ conjoin
+                        [ txBalance sealedTx combinedUTxO === mempty
+                        , (prop_expectFeeExcessSmallerThan
+                            (Cardano.Lovelace 1_200_000) sealedTx)
```
Perhaps it would be worth adding a comment here to describe the reasoning behind the 1.2 ada value?

Anviking replied:
2c533e2cd312ffc9e505a9be79ee5a7cdd55a49e

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2608,7 +2626,8 @@ prop_balanceTransactionBalanced (Wallet' utxo wal pending) (ShowBuildable partia
                 (SelectionLimitReached _)))) ->
-                label "selection limit reached" $ property True
+                    let msg = "selection limit reached should never be returned"
+                    in counterexample msg $ property False
```
👍🏻

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2617,7 +2636,75 @@ prop_balanceTransactionBalanced (Wallet' utxo wal pending) (ShowBuildable partia
+    walletUTxO :: UTxO
+    walletUTxO =
+          let Wallet' _ w _ = wallet
+          in view #utxo w
```
```suggestion
      let Wallet' _ w _ = wallet
      in view #utxo w
```

Anviking committed:
Reduce noise by moving deconstruction to helpers

Anviking committed:
Fixup: indentation

Anviking committed:
prop_balanceTransaction{Balanced -> Valid} + polish

- Rename the property to make it clear that it checks more things than
just whether the result is balanced.

- make the feeExcess expectation better motivated.

- polish a few other old comments

Anviking committed:
Add comment in genEncodingBoundaryLovelace

Anviking committed:
Improve note about bad input-resolution: ADP-1544

Anviking committed:
Update lib/core/src/Cardano/Wallet.hs

Anviking committed:
Add another note and link to ADP-1514

Anviking commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2546,21 +2544,20 @@ balanceTransactionGoldenSpec = describe "balance goldens" $ do
-prop_balanceTransactionBalanced wallet (ShowBuildable partialTx') seed
-    = withMaxSuccess 1000 $ do
+prop_balanceTransactionValid wallet (ShowBuildable partialTx') seed
+    = withMaxSuccess 20000 $ do
```
Oops. Didn't mean to commit the change to `20000`. Should revert

Anviking committed:
Add prop_posAndNegFromCardanoValueRoundtrip test

Anviking committed:
fixup: revert accidental increase to withMaxSuccess

Anviking committed:
clarify cbor ada bounds in comment

jonathanknowles committed:
Minor fixes to indendation.

- Use 4-space indents everywhere.
- Avoid vertical alignment.
- Ensure that case expressions and pattern guards are indented one level to
  the left of their handlers.

jonathanknowles committed:
Fix some spelling errors.

jonathanknowles committed:
Wrap long comments in `balanceTransactionWithSelectionStrategy`.

These were introduced in a prior PR, but we might as well clean them up
while we're adjusting this function.

https://input-output-hk.github.io/adrestia/code/Coding-Standards#limit-line-length-to-80-characters

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1524,43 +1535,83 @@ instance Buildable PartialTx where
+        -- 'assignScriptRedeemers'. This should make the problem (1) unlikely,
+        -- but not necessarily impossible.
+        --
+        -- We should investigate and make sure to handle this better.
```
👍🏻

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1570,49 +1621,45 @@ balanceTransaction
-                transform
+
+        lift $ traceWith tr $ MsgSelectionForBalancingStart
+            (CS.toExternalUTxOMap $ UTxOIndex.toMap internalUtxoAvailable)
```
I suspect this conversion can be avoided if we don't require ourselves to include a `UTxO` value in the `MsgSelectionForBalancingStart` value, since we only ever care about the size, which is the same before and after applying `toExternalUTxOMap`.

I'll push a commit that removes this conversion.

jonathanknowles replied:
Fixed in 80f346ce232504e4cf01094193e660e5646793ad.

jonathanknowles committed:
Remove unnecessary UTxO map conversions when constructing log values.

When recording details about the UTxO set in the log, we generally only
care about the size of the UTxO set, rather than its composition.

Therefore, we don't need to require ourselves to produce a value of type
`UTxO` in order to create a log entry: we just need to obtain the size
of the available UTxO set.

Therefore, we can avoid converting from the coin selection's internal
map of UTxO entries to values of the wallet's `UTxO` type, which
requires traversing the entire UTxO set and converting every single
entry.

We can't avoid traversing the wallet's UTxO set at least once, in order
to make a value of type `UTxOIndex`, but we can avoid doing it multiple
times.

jonathanknowles committed:
Replace `fromIntegral` with `intCastMaybe` in `posAndNegFromCardanoValue`.

This allows us to replace the pattern guard and the unsafe conversion
with a single checked conversion.

jonathanknowles reviewed (approved):
This looks reasonable to me. I see that you've highlighted all areas for further work with `TODO` comments and links to tickets, which is great.

Thanks for making this PR!

👍🏻
